stdlib/SparseArrays/src/linalg.jl

```diff
@@ -1017,11 +1017,33 @@ function rmul!(A::SparseMatrixCSC, b::Number)
     rmul!(A.nzval, b)
     return A
 end
+
 function lmul!(b::Number, A::SparseMatrixCSC)
     lmul!(b, A.nzval)
     return A
 end
 
+function rmul!(A::SparseMatrixCSC, D::Diagonal{T}) where T
+    m, n = size(A)
+    (n == size(D, 1)) || throw(DimensionMismatch())
+    Anzval = A.nzval
+    for col = 1:n, p = A.colptr[col]:(A.colptr[col+1]-1)
+        @inbounds Anzval[p] *= D.diag[col]
```

Comment:
It might be possible to gain a bit of speed here by manually hoisting the `D.diag[col]` lookup outside of the inner loop. Of course, the compiler may well be able to do that automatically. Depends on how much it knows about whether `D.diag` can change or move.

Reply:
From my experience with the SparseArray code, this is only needed when the object one `getfield` into is `mutable`. There is a lot of hoisting out fields for `SparseMatrixCSC` that was put there before https://github.com/JuliaLang/julia/pull/16371 (see discussion in https://github.com/JuliaLang/julia/issues/15668) but it shouldn't be needed now. Always worth benchmarking though...

Reply:
I'm sorry I can't follow. I did the change @StefanKarpinski suggested. Was that just (potentially) unnecessary or even "wrong"?

Reply:
I benchmarked the previous ("my") against the latest version ("Stefan's") on these:
```
P = sprand(12000, 12000, 0.01)
q = Diagonal(dropdims(sum(P, dims=2), dims=2))
```
and Stefan's won with `607.996 μs` vs. `631.496 μs`. So I guess it was worth it.

Reply:
The question is if the compiler tries to `reload `D.diag` every time in the loop or if it realizes that this will never change and hoist the load outside the loop.

Reply:
Seems like it doesn't reload, at least it doesn't show up in performance loss. Should I, for better readability, then merge the two loops again as before, but leave the `@inbounds` upfront like in the last commit?

Reply:
Doesn't matter that much but perhaps it makes sense to at least have the two methods in this PR be consistent with eachother.